### ISSUE TITLE
Bug Report: rpm build fails when attempting to generate tarball for development

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/vfatqc-python-scripts#168]
## Brief summary of issue

Following instructions noted [here](https://gist.github.com/jsturdy/4f9bccbdd18d953f73777e492a71fd24#file-dev-setup-md) had a build error when trying to build the rpm.

### Types of issue

- [X] Bug report (report an issue with the code)
- [ ] Feature request (request for change which adds functionality)

## Expected Behavior

I would be given a tarball to install in my `venv` so that I could test changes.

## Current Behavior

Build fails with error:

```
error: File not found by glob: /afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/vfatqc-python-scripts/rpm/build/bdist.linux-x86_64/rpm/BUILDROOT/gempython_vfatqc-2.0.0-centos7.python2.7.x86_64/usr/lib/python*/site-packages/gempython/scripts/*.py


RPM build errors:
    File not found by glob: /afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/vfatqc-python-scripts/rpm/build/bdist.linux-x86_64/rpm/BUILDROOT/gempython_vfatqc-2.0.0-centos7.python2.7.x86_64/usr/lib/python*/site-packages/gempython/scripts/*.py
error: command 'rpmbuild' failed with exit status 1
make: *** [_rpmbuild] Error 1
```

### Steps to Reproduce (for bugs)

1. Follow build instructions [here](https://gist.github.com/jsturdy/4f9bccbdd18d953f73777e492a71fd24#file-dev-setup-md)
2. `make && make rpm 2>&1 | tee vfatqc_rpm_build.txt`

I have attached the log file for the build output.
[vfatqc_rpm_build.txt](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/files/2059830/vfatqc_rpm_build.txt)

## Context (for feature requests)

Cannot make a tarball following the developer instructions.

## Your Environment

* Version used: 1c4b69b534a16536b920803c6f8aa6d5f7a6e10d
* Shell used: `/bin/zsh`

**Comment by jsturdy** [comment]
```zsh
git clone git@github.com:cms-gem-daq-project/vfatqc-python-scripts.git
cd vfatqc-python-scripts
make
make rpm
```
works for me on `gem904daq02`

```zsh
 git br -vv
* develop 1c4b69b [origin/develop] apply patch from #167
```

**Comment by bdorney** [comment]
>git clone git@github.com:cms-gem-daq-project/vfatqc-python-scripts.git
cd vfatqc-python-scripts
make
make rpm
works for me

Would the `$SHELL` env influence this at all?  e.g. did you try this on a fresh shell or did you execute some additional commands to setup the env?

**Comment by bdorney** [comment]
```
(py2.7)% git branch -vv      
* develop        1c4b69b [origin/develop] apply patch from #167
(py2.7)% make
(py2.7)% make rpm
RPM build errors:
    File not found by glob: /afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/vfatqc-python-scripts/rpm/build/bdist.linux-x86_64/rpm/BUILDROOT/gempython_vfatqc-2.0.0-centos7.python2.7.x86_64/usr/lib/python*/site-packages/gempython/scripts/*.py
error: command 'rpmbuild' failed with exit status 1
make: *** [_rpmbuild] Error 1
```

On `gem904qc8daq`.  I will try now on `gem904daq02`.

**Comment by jsturdy** [comment]
Things that might have an impact :
```zsh
pip list --user
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
cmsgemos-gempython (0.3.1)
numpy (1.14.2)
pbr (3.1.1)
root-numpy (4.7.3)
setuptools (39.0.1)
setuptools-scm (1.15.0)
wheel (0.31.0)
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```
But actually, I just tried inside a `virtualenv` and reproduced your issue.
This is because of the way the `virtualenv` is structured and what the `spec` file is assuming the filesystem will look like.
I will investigate ways to address this, but there's no *need* to do the `make` inside a `virtualenv`, that's only necessary for testing, so I will probably revise my statement to say "building can **not** be done *inside* the `virtualenv`"

**Comment by jsturdy** [comment]
Here's what the `BUILDROOT` directory structure looks like when building "normally"
```
/tmp/sturdy/vfatqc-python-scripts/rpm/build/bdist.linux-x86_64/rpm/BUILDROOT/gempython_vfatqc-2.0.0-centos7.python2.7.x86_64
├── opt
│   └── cmsgemos
│       └── bin
│           ├── confAllChambers.py
│           ├── confChamber.py
│           ├── fastLatency.py
│           ├── run_scans.py
│           ├── trimChamber.py
│           ├── ultraLatency.py
│           ├── ultraScurve.py
│           └── ultraThreshold.py
└── usr
    └── lib
        └── python2.7
            └── site-packages
                ├── gempython
                │   ├── scripts
```
And here's what it looks like *inside* the `virtualenv`
```
/tmp/sturdy/vfatqc-python-scripts/rpm/build/bdist.linux-x86_64/rpm/BUILDROOT/gempython_vfatqc-2.0.0-1.x86_64
├── opt
│   └── cmsgemos
│       └── bin
│           ├── confAllChambers.py
│           ├── confChamber.py
│           ├── fastLatency.py
│           ├── run_scans.py
│           ├── trimChamber.py
│           ├── ultraLatency.py
│           ├── ultraScurve.py
│           └── ultraThreshold.py
└── tmp
    └── sturdy
        └── vfatqc-python-scripts
            └── building
                └── lib
                    └── python2.7
                        └── site-packages
                            ├── gempython
                            │   ├── scripts
```

From this I hope it is clear that a releasable RPM **can not** (currently) be built from *inside* a `virtualenv`

**Comment by bdorney** [comment]
I agree there are differences; but it's my own ignorance as to how these differences cause this issue.

However:
> so I will probably revise my statement to say "building can not be done inside the virtualenv"

Seems like the critical take-away message.

**Comment by bdorney** [comment]
From a fresh shell, no `venv` on `gem904qc8daq`:

```
make
make rpm
Running _rpmbuild target
cd /afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/vfatqc-python-scripts/rpm && python setup.py bdist_rpm \
--release centos7.python2.7 \
--binary-only --force-arch=`uname -m`
Traceback (most recent call last):
  File "setup.py", line 56, in <module>
    license          = 'MIT',
  File "/usr/lib64/python2.7/distutils/core.py", line 112, in setup
    _setup_distribution = dist = klass(attrs)
  File "/usr/lib/python2.7/site-packages/setuptools/dist.py", line 265, in __init__
    self.fetch_build_eggs(attrs.pop('setup_requires'))
  File "/usr/lib/python2.7/site-packages/setuptools/dist.py", line 289, in fetch_build_eggs
    parse_requirements(requires), installer=self.fetch_build_egg
  File "/usr/lib/python2.7/site-packages/pkg_resources.py", line 630, in resolve
    raise VersionConflict(dist,req) # XXX put more info here
pkg_resources.VersionConflict: (setuptools 0.9.8 (/usr/lib/python2.7/site-packages), Requirement.parse('setuptools>=25.0'))
make: *** [_rpmbuild] Error 1
```

From a fresh shell, no `venv` on `gem904daq02`:

```
make 
make rpm
Running _rpmbuild target
cd /afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/vfatqc-python-scripts/rpm && python setup.py bdist_rpm \
--release centos7.python2.7 \
--binary-only --force-arch=`uname -m`
Traceback (most recent call last):
  File "setup.py", line 56, in <module>
    license          = 'MIT',
  File "/usr/lib64/python2.7/distutils/core.py", line 112, in setup
    _setup_distribution = dist = klass(attrs)
  File "/usr/lib/python2.7/site-packages/setuptools/dist.py", line 265, in __init__
    self.fetch_build_eggs(attrs.pop('setup_requires'))
  File "/usr/lib/python2.7/site-packages/setuptools/dist.py", line 289, in fetch_build_eggs
    parse_requirements(requires), installer=self.fetch_build_egg
  File "/usr/lib/python2.7/site-packages/pkg_resources.py", line 630, in resolve
    raise VersionConflict(dist,req) # XXX put more info here
pkg_resources.VersionConflict: (setuptools 0.9.8 (/usr/lib/python2.7/site-packages), Requirement.parse('setuptools>=25.0'))
make: *** [_rpmbuild] Error 1
```

**Comment by bdorney** [comment]
Perhaps some additional setup is required...?

**Comment by jsturdy** [comment]
Yes, the updates to `pip/setuptools/importlib` are needed (well, at the very least `importlib/setuptools`), and the only workaround is a `pip install --user`
However, if **not** building a releasable RPM, the `pip` installable zip can *probably* be built inside a `virtualenv` with an additional target in the `Makefile`, tested and now available in #171

**Comment by bdorney** [comment]
> Yes, the updates to pip/setuptools/importlib are needed (well, at the very least importlib/setuptools), and the only workaround is a pip install --user
However, if not building a releasable RPM, the pip installable zip can probably be built inside a virtualenv with an additional target in the Makefile, tested and now available in #171

Okay, I can try to do a pip install --user on importib and setuptools.  This should place theses packages in something like `$HOME/.local/../some/path`?  Then I just have to add `$HOME/.local/../some/path` to `$PYTHONPATH`?  Or should I wait for #171 to be merged in?

**Comment by bdorney** [comment]
Executing:

```
% pip install --user -I importlib 'setuptools<39.2.0'  'pip<10.1.0' 
% cd vfatqc-python-scripts
% make
% make rpm
```

Successfully built the rpm.  Thanks for feedback & support. Not sure if you want to close this issue or not.
